File: backend/src/main/java/com/agentplatform/backend/agents/AgentService.java
package com.agentplatform.backend.agents;

import com.agentplatform.backend.agents.dto.*;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class AgentService {

    private final AgentRepository agentRepository;
    private final AgentCapabilityRepository capabilityRepository;
    private final AgentToolPermissionRepository toolPermissionRepository;
    private final AgentPolicyRepository policyRepository;
    private final AgentRelationshipRepository relationshipRepository;

    public AgentService(AgentRepository agentRepository,
                        AgentCapabilityRepository capabilityRepository,
                        AgentToolPermissionRepository toolPermissionRepository,
                        AgentPolicyRepository policyRepository,
                        AgentRelationshipRepository relationshipRepository) {
        this.agentRepository = agentRepository;
        this.capabilityRepository = capabilityRepository;
        this.toolPermissionRepository = toolPermissionRepository;
        this.policyRepository = policyRepository;
        this.relationshipRepository = relationshipRepository;
    }

    // Agent CRUD
    public AgentResponse createAgent(Long ownerId, AgentCreateRequest req) {
        var a = new Agent(ownerId, req.getName(), req.getDescription());
        if (req.getTemplateId() != null) a.setTemplateId(req.getTemplateId());
        if (req.getModelPreference() != null) a.setModelPreference(req.getModelPreference());
        if (req.getInstructions() != null) a.setInstructions(req.getInstructions());
        if (req.getSystemPrompt() != null) a.setSystemPrompt(req.getSystemPrompt());
        var saved = agentRepository.save(a);
        
        // Create capabilities from wizard data
        if (req.getCapabilities() != null && !req.getCapabilities().isEmpty()) {
            for (String capName : req.getCapabilities()) {
                var cap = new AgentCapability(saved, capName, "Added from wizard");
                capabilityRepository.save(cap);
            }
        }
        
        // Create tool permissions from wizard data
        if (req.getAllowedTools() != null && !req.getAllowedTools().isEmpty()) {
            for (String toolName : req.getAllowedTools()) {
                var perm = new AgentToolPermission(saved, toolName, "execute", 
                    req.getApproveEveryAction() != null ? req.getApproveEveryAction() : false);
                toolPermissionRepository.save(perm);
            }
        }
        
        // Create policies from wizard data
        if (req.getApproveEveryAction() != null && req.getApproveEveryAction()) {
            var policy = new AgentPolicy(saved, "approveEveryAction", "true");
            policyRepository.save(policy);
        }
        if (req.getRememberApprovals() != null && req.getRememberApprovals()) {
            var policy = new AgentPolicy(saved, "rememberApprovals", "true");
            policyRepository.save(policy);
        }
        if (req.getEnableMemory() != null) {
            var policy = new AgentPolicy(saved, "enableMemory", req.getEnableMemory().toString());
            policyRepository.save(policy);
        }
        if (req.getMemoryType() != null) {
            var policy = new AgentPolicy(saved, "memoryType", req.getMemoryType());
            policyRepository.save(policy);
        }
        
        return map(saved);
    }

    public List<AgentResponse> listByOwner(Long ownerId) {
        return agentRepository.findByOwnerId(ownerId).stream().map(this::map).collect(Collectors.toList());
    }

    public AgentResponse getByIdForOwner(Long ownerId, Long agentId) {
        var opt = agentRepository.findById(agentId);
        if (opt.isEmpty() || !opt.get().getOwnerId().equals(ownerId)) return null;
        return map(opt.get());
    }

    public void deleteByIdForOwner(Long ownerId, Long agentId) {
        var opt = agentRepository.findById(agentId);
        if (opt.isPresent() && opt.get().getOwnerId().equals(ownerId)) {
            agentRepository.deleteById(agentId);
        }
    }

    public AgentResponse updateAgent(Long ownerId, Long agentId, AgentCreateRequest req) {
        var opt = agentRepository.findById(agentId);
        if (opt.isEmpty() || !opt.get().getOwnerId().equals(ownerId)) return null;
        var a = opt.get();
        if (req.getName() != null) a.setName(req.getName());
        if (req.getDescription() != null) a.setDescription(req.getDescription());
        if (req.getTemplateId() != null) a.setTemplateId(req.getTemplateId());
        if (req.getModelPreference() != null) a.setModelPreference(req.getModelPreference());
        if (req.getInstructions() != null) a.setInstructions(req.getInstructions());
        if (req.getSystemPrompt() != null) a.setSystemPrompt(req.getSystemPrompt());
        a.setUpdatedAt(Instant.now());
        var saved = agentRepository.save(a);
        return map(saved);
    }

    // Capability management
    public AgentCapabilityResponse addCapability(Long ownerId, Long agentId, String name, String description) {
        var opt = agentRepository.findById(agentId);
        if (opt.isEmpty() || !opt.get().getOwnerId().equals(ownerId)) return null;
        var cap = new AgentCapability(opt.get(), name, description);
        var saved = capabilityRepository.save(cap);
        return mapCapability(saved);
    }

    public List<AgentCapabilityResponse> listCapabilities(Long ownerId, Long agentId) {
        var opt = agentRepository.findById(agentId);
        if (opt.isEmpty() || !opt.get().getOwnerId().equals(ownerId)) return List.of();
        return capabilityRepository.findByAgent(opt.get()).stream()
                .map(this::mapCapability).collect(Collectors.toList());
    }

    public void removeCapability(Long ownerId, Long agentId, Long capabilityId) {
        var opt = agentRepository.findById(agentId);
        if (opt.isPresent() && opt.get().getOwnerId().equals(ownerId)) {
            capabilityRepository.deleteById(capabilityId);
        }
    }

    // Tool permission management
    public AgentToolPermissionResponse addToolPermission(Long ownerId, Long agentId, String toolName, String permLevel, Boolean requiresApproval) {
        var opt = agentRepository.findById(agentId);
        if (opt.isEmpty() || !opt.get().getOwnerId().equals(ownerId)) return null;
        var perm = new AgentToolPermission(opt.get(), toolName, permLevel, requiresApproval);
        var saved = toolPermissionRepository.save(perm);
        return mapToolPermission(saved);
    }

    public List<AgentToolPermissionResponse> listToolPermissions(Long ownerId, Long agentId) {
        var opt = agentRepository.findById(agentId);
        if (opt.isEmpty() || !opt.get().getOwnerId().equals(ownerId)) return List.of();
        return toolPermissionRepository.findByAgent(opt.get()).stream()
                .map(this::mapToolPermission).collect(Collectors.toList());
    }

    // Policy management
    public AgentPolicyResponse addPolicy(Long ownerId, Long agentId, String policyName, String policyValue) {
        var opt = agentRepository.findById(agentId);
        if (opt.isEmpty() || !opt.get().getOwnerId().equals(ownerId)) return null;
        var policy = new AgentPolicy(opt.get(), policyName, policyValue);
        var saved = policyRepository.save(policy);
        return mapPolicy(saved);
    }

    public List<AgentPolicyResponse> listPolicies(Long ownerId, Long agentId) {
        var opt = agentRepository.findById(agentId);
        if (opt.isEmpty() || !opt.get().getOwnerId().equals(ownerId)) return List.of();
        return policyRepository.findByAgent(opt.get()).stream()
                .map(this::mapPolicy).collect(Collectors.toList());
    }

    // Mapping functions
    private AgentResponse map(Agent a) {
        var resp = new AgentResponse(a.getId(), a.getOwnerId(), a.getName(), a.getDescription(), a.getStatus(), a.getVisibility(), a.getCreatedAt());
        resp.setTemplateId(a.getTemplateId());
        resp.setModelPreference(a.getModelPreference());
        resp.setInstructions(a.getInstructions());
        resp.setSystemPrompt(a.getSystemPrompt());
        resp.setUpdatedAt(a.getUpdatedAt());
        
        if (a.getCapabilities() != null) {
            resp.setCapabilities(a.getCapabilities().stream().map(this::mapCapability).collect(Collectors.toList()));
        }
        if (a.getToolPermissions() != null) {
            resp.setToolPermissions(a.getToolPermissions().stream().map(this::mapToolPermission).collect(Collectors.toList()));
        }
        if (a.getPolicies() != null) {
            resp.setPolicies(a.getPolicies().stream().map(this::mapPolicy).collect(Collectors.toList()));
        }
        
        return resp;
    }

    private AgentCapabilityResponse mapCapability(AgentCapability c) {
        return new AgentCapabilityResponse(c.getId(), c.getCapabilityName(), c.getDescription(), c.getEnabled(), c.getCreatedAt());
    }

    private AgentToolPermissionResponse mapToolPermission(AgentToolPermission p) {
        return new AgentToolPermissionResponse(p.getId(), p.getToolName(), p.getPermissionLevel(), p.getRequiresApproval(), p.getCreatedAt());
    }

    private AgentPolicyResponse mapPolicy(AgentPolicy p) {
        return new AgentPolicyResponse(p.getId(), p.getPolicyName(), p.getPolicyValue(), p.getCreatedAt());
    }
}

File: frontend/agentplatform/src/app/services/agent.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface AgentCapabilityResponse {
  id: number;
  capabilityName: string;
  description?: string;
  enabled: boolean;
  createdAt?: string;
}

export interface AgentToolPermissionResponse {
  id: number;
  toolName: string;
  permissionLevel: string;
  requiresApproval: boolean;
  createdAt?: string;
}

export interface AgentPolicyResponse {
  id: number;
  policyName: string;
  policyValue: string;
  createdAt?: string;
}

export interface AgentResponse {
  id: number;
  ownerId?: number;
  templateId?: number;
  name: string;
  description?: string;
  status?: string;
  visibility?: string;
  modelPreference?: string;
  instructions?: string;
  systemPrompt?: string;
  capabilities?: AgentCapabilityResponse[];
  toolPermissions?: AgentToolPermissionResponse[];
  policies?: AgentPolicyResponse[];
  createdAt?: string;
  updatedAt?: string;
}

export interface ParseAgentRequest {
  prompt: string;
}

export interface ProposedAgent {
  name: string;
  role: string;
  capabilities: string[];
  toolPermissions: string[];
  approvalSummary: string;
}

export interface ParseAgentResponse {
  proposedAgents: ProposedAgent[];
  requiresReview: boolean;
}

@Injectable({ providedIn: 'root' })
export class AgentService {
  private readonly baseUrl = 'http://localhost:8083';

  constructor(private http: HttpClient) {}

  // Agent CRUD
  getAgents(): Observable<AgentResponse[]> {
    return this.http.get<AgentResponse[]>(`${this.baseUrl}/api/agents`);
  }

  create(req: any): Observable<AgentResponse> {
    return this.http.post<AgentResponse>(`${this.baseUrl}/api/agents`, req);
  }

  getAgent(id: number): Observable<AgentResponse> {
    return this.http.get<AgentResponse>(`${this.baseUrl}/api/agents/${id}`);
  }

  updateAgent(id: number, req: { name?: string; description?: string; templateId?: number; modelPreference?: string; instructions?: string; systemPrompt?: string }): Observable<AgentResponse> {
    return this.http.put<AgentResponse>(`${this.baseUrl}/api/agents/${id}`, req);
  }

  deleteAgent(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/api/agents/${id}`);
  }

  // Capability management
  getCapabilities(agentId: number): Observable<AgentCapabilityResponse[]> {
    return this.http.get<AgentCapabilityResponse[]>(`${this.baseUrl}/api/agents/${agentId}/capabilities`);
  }

  addCapability(agentId: number, capabilityName: string, description?: string): Observable<AgentCapabilityResponse> {
    return this.http.post<AgentCapabilityResponse>(`${this.baseUrl}/api/agents/${agentId}/capabilities`,
      { capabilityName, description });
  }

  removeCapability(agentId: number, capabilityId: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/api/agents/${agentId}/capabilities/${capabilityId}`);
  }

  // Tool permission management
  getToolPermissions(agentId: number): Observable<AgentToolPermissionResponse[]> {
    return this.http.get<AgentToolPermissionResponse[]>(`${this.baseUrl}/api/agents/${agentId}/tool-permissions`);
  }

  addToolPermission(agentId: number, toolName: string, permissionLevel?: string, requiresApproval?: boolean): Observable<AgentToolPermissionResponse> {
    return this.http.post<AgentToolPermissionResponse>(`${this.baseUrl}/api/agents/${agentId}/tool-permissions`,
      { toolName, permissionLevel: permissionLevel || 'execute', requiresApproval: requiresApproval || false });
  }

  // Policy management
  getPolicies(agentId: number): Observable<AgentPolicyResponse[]> {
    return this.http.get<AgentPolicyResponse[]>(`${this.baseUrl}/api/agents/${agentId}/policies`);
  }

  addPolicy(agentId: number, policyName: string, policyValue: string): Observable<AgentPolicyResponse> {
    return this.http.post<AgentPolicyResponse>(`${this.baseUrl}/api/agents/${agentId}/policies`,
      { policyName, policyValue });
  }

  // Parser
  parseDescription(req: ParseAgentRequest): Observable<ParseAgentResponse> {
    return this.http.post<ParseAgentResponse>(`${this.baseUrl}/api/agents/parse`, req);
  }
}
